ASFAchievementManager/AchievementHandler.cs

```diff
@@ -21,53 +19,52 @@ public sealed class AchievementHandler : ClientMsgHandler {
 
 			switch (packetMsg.MsgType) {
 				case EMsg.ClientGetUserStatsResponse:
-					HandleGetUserStatsResponse(packetMsg);
-
+					ClientMsgProtobuf<CMsgClientGetUserStatsResponse> getAchievementsResponse = new(packetMsg);
+					Client.PostCallback(new GetAchievementsCallback(packetMsg.TargetJobID, getAchievementsResponse.Body));
 					break;
 				case EMsg.ClientStoreUserStatsResponse:
-					HandleStoreUserStatsResponse(packetMsg);
-
+					ClientMsgProtobuf<CMsgClientStoreUserStatsResponse> setAchievementsResponse = new(packetMsg);
+					Client.PostCallback(new SetAchievementsCallback(packetMsg.TargetJobID, setAchievementsResponse.Body));
 					break;
 			}
 
 		}
 
-		private void HandleGetUserStatsResponse(IPacketMsg packetMsg) {
-			if (packetMsg == null) {
-				ASF.ArchiLogger.LogNullError(nameof(packetMsg));
-				return;
-			}
-			ClientMsgProtobuf<CMsgClientGetUserStatsResponse> response = new ClientMsgProtobuf<CMsgClientGetUserStatsResponse>(packetMsg);
-			if (!Responses.TryAdd(response.Body.game_id, new StoredResponse {
-				Success = response.Body.eresult == 1,
-				Response = response.Body
-			})) {
-				ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, "GetAchievements"));
-			}
+		internal abstract class AchievementsCallBack<T> : CallbackMsg {
+			internal readonly T Response;
+			internal readonly bool Success;
 
-			if (response.Body.eresult != 1) {
-				ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, "GetAchievements"));
-			}
+			internal AchievementsCallBack(JobID jobID, T msg, Func<T, EResult> eresultGetter, string error) {
+				if (jobID == null) {
+					throw new ArgumentNullException(nameof(jobID));
+				}
 
-		}
+				if (msg == null) {
+					throw new ArgumentNullException(nameof(msg));
+				}
 
-		private void HandleStoreUserStatsResponse(IPacketMsg packetMsg) {
-			if (packetMsg == null) {
-				ASF.ArchiLogger.LogNullError(nameof(packetMsg));
-				return;
-			}
-			ClientMsgProtobuf<CMsgClientStoreUserStatsResponse> response = new ClientMsgProtobuf<CMsgClientStoreUserStatsResponse>(packetMsg);
-			if (!Responses.TryAdd(response.Body.game_id, new StoredResponse {
-				Success = response.Body.eresult == 1,
-				Response = null //we don't care about this, just need to know that request was successful
-			})) {
-				ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, "SetAchievements"));
+				JobID = jobID;
+				Success = eresultGetter(msg) == EResult.OK;
+				Response = msg;
+
+				if (!Success) {
+					LogFailure(error);
+				}
 			}
 
-			if (response.Body.eresult != 1) {
-				ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, "SetAchievements"));
+			protected void LogFailure(string error) {
```

Comment:
Not used in derived class and a single line of code. I personally would insert this single statement at Line 51 to not complicate everything here. Especially since the compiler will inline this statement in a non-debug build anyways.